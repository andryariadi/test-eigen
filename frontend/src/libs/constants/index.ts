export const headlinesCategories = ["general", "business", "entertainment", "health", "science", "sports", "technology"];

export const everythingLanguages = ["ar", "de", "en", "es", "fr", "he", "it", "nl", "no", "pt", "ru", "sv", "ud", "zh"];

export const newsSelection = ["everything", "top-headlines"];
