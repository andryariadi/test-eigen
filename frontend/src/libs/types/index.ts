interface ArticleSource {
  id: string | null;
  name: string;
}

export interface ArticleProps {
  source: ArticleSource;
  author: string | null;
  title: string;
  description: string | null;
  url: string;
  urlToImage: string | null;
  publishedAt: string;
  content: string | null;
}
