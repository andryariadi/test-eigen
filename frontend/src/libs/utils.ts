import { clsx, type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function formatNewsDate(isoDate: string): string {
  const date = new Date(isoDate);

  const options: Intl.DateTimeFormatOptions = {
    month: "long",
    day: "numeric",
    year: "numeric",
  };

  return date.toLocaleDateString("en-US", options);
}

export const stripHtmlTags = (html: string | null) => {
  if (typeof window === "undefined") {
    // Untuk server-side
    return html ? html.replace(/<[^>]*>?/gm, "") : "";
  } else {
    // Untuk client-side
    const doc = new DOMParser().parseFromString(html ?? "", "text/html");
    return doc.body.textContent || "";
  }
};
