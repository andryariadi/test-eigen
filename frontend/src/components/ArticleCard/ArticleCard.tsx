import { ArticleProps } from "@/libs/types";
import { formatNewsDate, stripHtmlTags } from "@/libs/utils";
import Image from "next/image";
import Link from "next/link";
import React from "react";

const ArticleCard = ({ article }: { article: ArticleProps }) => {
  //   console.log({ article }, "<---articleCard");

  return (
    <figure className="group h-[423px] hover:bg-white hover:shadow-lg trassation-all duration-300 overflow-hidden rounded-[12px]">
      {/* Image */}
      <Link href={`/article/${article.source.name}`}>
        <div className="b-sky-800 relative h-full max-h-[240px] w-full overflow-hidden rounded-[12px]">
          <Image src={article.urlToImage ?? "https://placehold.co/600x400/png"} fill alt={article.title} className="object-cover" />
        </div>
      </Link>

      <figcaption className="b-amber-600 mt-4 flex flex-col gap-2 group-hover:px-2 transition-all duration-300 text-slate-600">
        {/* Date */}
        <span className="text-sm font-normal">{formatNewsDate(article.publishedAt)}</span>

        {/* Title */}
        <Link href={`/article/${article.source.name}`}>
          <h2 className="text-lg text-slate-900 font-semibold line-clamp-1 capitalize">{article.title}</h2>
        </Link>

        {/* Description */}
        <p className="text-base font-normal line-clamp-2">{stripHtmlTags(article.description)}</p>

        {/* Category */}
        <div className="bg-blue-200 w-max px-3 py-1 rounded-full">
          <Link href={article.url} target="_blank" className="text-blue-900 text-sm font-normal">
            {article.source.name}
          </Link>
        </div>
      </figcaption>
    </figure>
  );
};

export default ArticleCard;
