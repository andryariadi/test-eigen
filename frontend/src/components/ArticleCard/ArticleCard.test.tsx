// ArticleCard.test.tsx
import { render, screen } from "@testing-library/react";
import ArticleCard from "./ArticleCard";
import { ArticleProps } from "@/libs/types";
import { formatNewsDate, stripHtmlTags } from "@/libs/utils";

// Mock Next.js components and utils
jest.mock("next/image", () => ({
  __esModule: true,
  default: ({ src, alt, className }: { src: string; alt: string; className: string }) => <img src={src} alt={alt} className={className} data-testid="article-image" />,
}));

jest.mock("next/link", () => ({
  __esModule: true,
  default: ({ children, href, ...props }: { children: React.ReactNode; href: string }) => (
    <a href={href} {...props}>
      {children}
    </a>
  ),
}));

jest.mock("../../libs/utils", () => ({
  __esModule: true,
  formatNewsDate: jest.fn((date: string) => `Formatted: ${date}`),
  stripHtmlTags: jest.fn((text: string | null) => (text ? text.replace(/<[^>]*>/g, "") : "")),
}));

describe("ArticleCard Component", () => {
  const mockArticle: ArticleProps = {
    source: {
      id: "1",
      name: "Tech News",
    },
    author: "John Doe",
    title: "React Testing in 2024",
    description: "<p>Learn how to test React components</p>",
    url: "https://example.com/article",
    urlToImage: "https://example.com/image.jpg",
    publishedAt: "2024-01-01T00:00:00Z",
    content: "Lorem ipsum dolor sit amet...",
  };

  beforeEach(() => {
    // Reset mocks before each test
    jest.clearAllMocks();
  });

  it("renders correctly with all article data", () => {
    render(<ArticleCard article={mockArticle} />);

    // 1. Test images
    const image = screen.getByTestId("article-image");

    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute("src", mockArticle.urlToImage);
    expect(image).toHaveAttribute("alt", mockArticle.title);
    expect(image).toHaveClass("object-cover");

    // 2. Test date format
    expect(formatNewsDate).toHaveBeenCalledWith(mockArticle.publishedAt);
    expect(screen.getByText(`Formatted: ${mockArticle.publishedAt}`)).toBeInTheDocument();

    // 3. Test link to article
    const titleLink = screen.getByTestId("article-title-link");

    expect(titleLink).toBeInTheDocument();
    expect(titleLink).toHaveAttribute("href", `/article/${mockArticle.source.name}`);

    // 4. Test article title
    const titleText = screen.getByText(mockArticle.title);

    expect(titleText).toHaveClass("capitalize");
    expect(titleText).toHaveClass("line-clamp-1");
    expect(titleText).toHaveClass("text-lg");
    expect(titleText).toHaveClass("text-slate-900");
    expect(titleText).toHaveClass("font-semibold");

    // 5. Test description
    expect(stripHtmlTags).toHaveBeenCalledWith(mockArticle.description);
    expect(screen.getByText("Learn how to test React components")).toBeInTheDocument();
    expect(screen.getByText("Learn how to test React components")).toHaveClass("line-clamp-2");

    // 6. Test category
    const categoryLink = screen.getByRole("link", { name: mockArticle.source.name });

    expect(categoryLink).toBeInTheDocument();
    expect(categoryLink).toHaveAttribute("href", mockArticle.url);
    expect(categoryLink).toHaveAttribute("target", "_blank");
    expect(categoryLink).toHaveClass("text-blue-900");
  });

  it("uses placeholder image when urlToImage is null", () => {
    const articleWithoutImage = {
      ...mockArticle,
      urlToImage: null,
    };
    render(<ArticleCard article={articleWithoutImage} />);

    expect(screen.getByTestId("article-image")).toHaveAttribute("src", "https://placehold.co/600x400/png");
  });

  it("handles empty description safely", () => {
    const articleWithoutDesc = {
      ...mockArticle,
      description: null,
    };
    render(<ArticleCard article={articleWithoutDesc} />);

    expect(stripHtmlTags).toHaveBeenCalledWith(null);
    expect(screen.queryByTestId("article-description")).toBeNull();
  });

  it("applies correct hover styles", () => {
    render(<ArticleCard article={mockArticle} />);

    const card = screen.getByRole("figure");

    expect(card).toHaveClass("group");
    expect(card).toHaveClass("hover:bg-white");
    expect(card).toHaveClass("hover:shadow-lg");
    expect(card).toHaveClass("transition-all");

    const caption = screen.getByTestId("figcaption");

    expect(caption).toHaveClass("group-hover:px-2");
  });

  it("matches snapshot with complete article data", () => {
    const { asFragment } = render(<ArticleCard article={mockArticle} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot with minimal article data", () => {
    const minimalArticle = {
      source: {
        id: null,
        name: "Minimal News",
      },
      title: "Minimal Article",
      url: "https://example.com/minimal",
      publishedAt: "2024-01-02T00:00:00Z",
    } as ArticleProps;

    const { asFragment } = render(<ArticleCard article={minimalArticle} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
