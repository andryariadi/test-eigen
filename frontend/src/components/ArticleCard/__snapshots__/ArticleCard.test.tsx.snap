// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ArticleCard Component matches snapshot with complete article data 1`] = `
<DocumentFragment>
  <figure
    class="group h-[423px] hover:bg-white hover:shadow-lg transition-all duration-300 overflow-hidden rounded-[12px]"
  >
    <a
      href="/article/Tech News/?"
    >
      <div
        class="b-sky-800 relative h-full max-h-[240px] w-full overflow-hidden rounded-[12px]"
      >
        <img
          alt="React Testing in 2024"
          class="object-cover"
          data-testid="article-image"
          src="https://example.com/image.jpg"
        />
      </div>
    </a>
    <figcaption
      class="b-amber-600 mt-4 flex flex-col gap-2 group-hover:px-2 transition-all duration-300 text-slate-600"
      data-testid="figcaption"
    >
      <span
        class="text-sm font-normal"
      >
        Formatted: 2024-01-01T00:00:00Z
      </span>
      <a
        data-testid="article-title-link"
        href="/article/Tech News/?"
      >
        <h2
          class="text-lg text-slate-900 font-semibold line-clamp-1 capitalize"
        >
          React Testing in 2024
        </h2>
      </a>
      <p
        class="text-base font-normal line-clamp-2"
      >
        Learn how to test React components
      </p>
      <div
        class="bg-blue-200 w-max px-3 py-1 rounded-full"
      >
        <a
          class="text-blue-900 text-sm font-normal"
          href="https://example.com/article"
          target="_blank"
        >
          Tech News
        </a>
      </div>
    </figcaption>
  </figure>
</DocumentFragment>
`;

exports[`ArticleCard Component matches snapshot with minimal article data 1`] = `
<DocumentFragment>
  <figure
    class="group h-[423px] hover:bg-white hover:shadow-lg transition-all duration-300 overflow-hidden rounded-[12px]"
  >
    <a
      href="/article/Minimal News/?"
    >
      <div
        class="b-sky-800 relative h-full max-h-[240px] w-full overflow-hidden rounded-[12px]"
      >
        <img
          alt="Minimal Article"
          class="object-cover"
          data-testid="article-image"
          src="https://placehold.co/600x400/png"
        />
      </div>
    </a>
    <figcaption
      class="b-amber-600 mt-4 flex flex-col gap-2 group-hover:px-2 transition-all duration-300 text-slate-600"
      data-testid="figcaption"
    >
      <span
        class="text-sm font-normal"
      >
        Formatted: 2024-01-02T00:00:00Z
      </span>
      <a
        data-testid="article-title-link"
        href="/article/Minimal News/?"
      >
        <h2
          class="text-lg text-slate-900 font-semibold line-clamp-1 capitalize"
        >
          Minimal Article
        </h2>
      </a>
      <p
        class="text-base font-normal line-clamp-2"
      />
      <div
        class="bg-blue-200 w-max px-3 py-1 rounded-full"
      >
        <a
          class="text-blue-900 text-sm font-normal"
          href="https://example.com/minimal"
          target="_blank"
        >
          Minimal News
        </a>
      </div>
    </figcaption>
  </figure>
</DocumentFragment>
`;
